### PR TITLE
[CUDA][Linalg} Patch crash of `linalg.eigh` when input matrix is ill-conditioned, in some cusolver version

### PR DESCRIPTION
**Pull request by xwang233**
Related: https://github.com/pytorch/pytorch/issues/94772, https://github.com/pytorch/pytorch/issues/105359

I can locally reproduce this crash with pytorch 2.0.1 stable pip binary. The test already passes with the latest cuda 12.2 release.

Re: https://github.com/pytorch/pytorch/issues/94772#issuecomment-1658909998
> From discussion in triage review:

- [x] we should add a test to prevent regressions
- [x] properly document support wrt different CUDA versions
- [x] possibly add support using MAGMA

### CODE REVIEW

**Comment by xwang233** [comment]
cc @ptrblck @malfet @soulitzer

**Comment by xwang233** [comment]
Seems like the test fails (as expected?) in cuda 12.1 CI. 😟  https://hud.pytorch.org/pytorch/pytorch/pull/107082?sha=829f99fa6412a858034db06b6894d680517ad019

```
2023-08-15T01:31:41.1333178Z _ TestLinalgCUDA.test_eigh_svd_illcondition_matrix_input_should_not_crash_cuda_float64 _
2023-08-15T01:31:41.1333576Z Traceback (most recent call last):
2023-08-15T01:31:41.1334007Z   File "/var/lib/jenkins/workspace/test/test_linalg.py", line 1002, in test_eigh_svd_illcondition_matrix_input_should_not_crash
2023-08-15T01:31:41.1334437Z     eigh_out = torch.linalg.eigh(a)
2023-08-15T01:31:41.1335560Z RuntimeError: cusolver error: CUSOLVER_STATUS_EXECUTION_FAILED, when calling `cusolverDnXsyevd( handle, params, jobz, uplo, n, CUDA_R_64F, reinterpret_cast<void*>(A), lda, CUDA_R_64F, reinterpret_cast<void*>(W), CUDA_R_64F, reinterpret_cast<void*>(bufferOnDevice), workspaceInBytesOnDevice, reinterpret_cast<void*>(bufferOnHost), workspaceInBytesOnHost, info)`. If you keep seeing this error, you may use `torch.backends.cuda.preferred_linalg_library()` to try linear algebra operators with other supported backends. See https://pytorch.org/docs/stable/backends.html#torch.backends.cuda.preferred_linalg_library
2023-08-15T01:31:41.1336507Z 
2023-08-15T01:31:41.1336673Z To execute this test, run the following from the base repo dir:
2023-08-15T01:31:41.1337330Z      python test/test_linalg.py -k test_eigh_svd_illcondition_matrix_input_should_not_crash_cuda_float64
2023-08-15T01:31:41.1337838Z 
2023-08-15T01:31:41.1338116Z This message can be suppressed by setting PYTORCH_PRINT_REPRO_ON_FAILURE=0
```

File: torch/linalg/__init__.py [hunk 1]
```diff
@@ -647,6 +647,13 @@
              :math:`\lambda_i` through the computation of
              :math:`\frac{1}{\min_{i \neq j} \lambda_i - \lambda_j}`.
 
+.. warning:: User may see pytorch crashes if running `eigh` on CUDA devices with CUDA versions before 12.1 update 1
+             with large ill-conditioned matrices as inputs.
+             Refer to :ref:`Linear Algebra Numerical Stability<Linear Algebra Stability>` for more details.
+             If this is the case, user may (1) tune their matrix inputs to be less ill-conditioned,
+             or (2) use :func:`torch.backends.cuda.preferred_linalg_library` to
+             try linear algebra operators with other supported backends.
```
Comment (lezcano) [inline torch/linalg/__init__.py:655]: nit
```suggestion
             try other supported backends.
```

File: test/test_linalg.py [hunk 2]
```diff
@@ -988,6 +988,26 @@ def test_eigh_errors_and_warnings(self, device, dtype):
             with self.assertRaisesRegex(RuntimeError, "tensors to be on the same device"):
                 torch.linalg.eigh(a, out=(out_w, out_v))
 
+    @skipCPUIfNoLapack
+    @dtypes(torch.float, torch.double)
+    @precisionOverride({torch.float32: 1, torch.float64: 1})
```
Comment (lezcano) [inline test/test_linalg.py:993]: Might make more sense to put some sane relative errors, as we are comparing rather large numbers.

Comment (xwang233) [reply test/test_linalg.py:993]: I thought precisionOverride is atol only? Let me check the code.

**Review by lezcano — APPROVED** [review]
If the tests pass this LGTM. A high tolerance as the test has LGTM, as the inputs have plenty of repeated eigenvalues.

**Pull request closed without merging**